### PR TITLE
Raise TypeError when scalar column added to empty table

### LINKED ISSUE #3811: Better error if adding a non-iterable object to a table as a column
In the following example, it would be nice to have a more explicit error:

```
In [2]: t = Table()

In [3]: t['a'] = 1
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-48681fe373f1> in <module>()
----> 1 t['a'] = 1

/Users/tom/miniconda3/envs/production/lib/python3.4/site-packages/astropy/table/table.py in __setitem__(self, item, value)
    909 
    910             # Now add new column to the table
--> 911             self.add_columns([new_column], copy=False)
    912 
    913         else:

/Users/tom/miniconda3/envs/production/lib/python3.4/site-packages/astropy/table/table.py in add_columns(self, cols, indexes, copy)
   1203                 newcols.insert(i, col)
   1204 
-> 1205         self._init_from_cols(newcols)
   1206 
   1207     def remove_row(self, index):

/Users/tom/miniconda3/envs/production/lib/python3.4/site-packages/astropy/table/table.py in _init_from_cols(self, cols)
    516         """Initialize table from a list of Column objects"""
    517 
--> 518         lengths = set(len(col) for col in cols)
    519         if len(lengths) != 1:
    520             raise ValueError('Inconsistent data column lengths: {0}'

/Users/tom/miniconda3/envs/production/lib/python3.4/site-packages/astropy/table/table.py in <genexpr>(.0)
    516         """Initialize table from a list of Column objects"""
    517 
--> 518         lengths = set(len(col) for col in cols)
    519         if len(lengths) != 1:
    520             raise ValueError('Inconsistent data column lengths: {0}'

TypeError: len() of unsized object
```

Passing a scalar value can happen accidentally, and it would be nice to have something saying that table columns can only be set to iterable objects.

@taldcroft - what do you think?

### PR DESCRIPTION
**Pull request by gregosimo**
### Description

This pull request is to address the uninformative error message when a column in an empty table is assigned to a scalar. Now assigning a column in an empty table to a scalar will result in an TypeError at assignment with a relevant error message, rather than a TypeError on access with a more opaque error message.

Fixes #3811

### CODE REVIEW

File: CHANGES.rst
Comment (pllim) [inline CHANGES.rst:936]:
```suggestion
- Raise a TypeError when a scalar column is added to an unsized table. [#3811, #10476]
```

File: astropy/table/table.py
Comment (taldcroft) [inline astropy/table/table.py:1952]:
```suggestion
            raise TypeError('Empty table cannot have column set to scalar value')
```

File: astropy/table/tests/test_table.py
Comment (taldcroft) [inline astropy/table/tests/test_table.py:263]:
```suggestion
        with pytest.raises(TypeError, match='Empty table cannot have column set to scalar value'):
```

File: CHANGES.rst
Comment (taldcroft) [inline CHANGES.rst:938]:
```suggestion
```

**Comment by taldcroft** [comment]
@gregosimo - looks good, thanks!  I made a few tiny tweaks and will merge when this passes CI.

**Comment by taldcroft** [comment]
I see now that the changelog entry is in the wrong section.  It is currently in 4.0.1 but should be in 4.2.

**Review by bsipocz — APPROVED** [review]
I take the discussion as an approval from the package maintainer, so let's just get this merged now.

**Comment by bsipocz** [comment]
Thank you @gregosimo!

**Pull request merged**
